common/persistence/visibility/visibility_manager_impl.go

```diff
@@ -152,49 +160,109 @@ func (p *visibilityManagerImpl) ListChasmExecutions(
+	}
+	chasmAliasedSAs, err := aliasChasmSearchAttributes(decodedChasmSAs, mapper)
+	if err != nil {
+		return nil, err
+	}
+
+	customAliasedSAs, err := searchattribute.AliasFields(
+		p.searchAttributesMapperProvider,
+		customSAs,
+		namespaceName.String(),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	userMemo, chasmMemoPayload, err := p.splitUserAndChasmMemo(exec)
+	if err != nil {
+		return nil, err
+	}
+
+	return &chasm.ExecutionInfo[*commonpb.Payload]{
+		BusinessID:             exec.WorkflowID,
+		RunID:                  exec.RunID,
+		StartTime:              exec.StartTime,
+		CloseTime:              exec.CloseTime,
+		HistoryLength:          exec.HistoryLength,
+		HistorySizeBytes:       exec.HistorySizeBytes,
+		StateTransitionCount:   exec.StateTransitionCount,
+		ChasmSearchAttributes:  chasm.NewSearchAttributesMap(chasmAliasedSAs),
+		CustomSearchAttributes: customAliasedSAs.GetIndexedFields(),
+		Memo:                   userMemo,
+		ChasmMemo:              chasmMemoPayload,
+	}, nil
+}
+
+// splitSearchAttributes splits decoded search attributes into CHASM and custom attributes.
+func splitSearchAttributes(searchAttributes *commonpb.SearchAttributes) (customSAs, chasmSAs *commonpb.SearchAttributes) {
+	chasmSAs = &commonpb.SearchAttributes{IndexedFields: make(map[string]*commonpb.Payload)}
+	customSAs = &commonpb.SearchAttributes{IndexedFields: make(map[string]*commonpb.Payload)}
+	for name, payloadValue := range searchAttributes.GetIndexedFields() {
+		if sadefs.IsChasmSearchAttribute(name) {
+			chasmSAs.IndexedFields[name] = payloadValue
 		} else {
-			// Archetype doesn't match or no combined memo - return entire memo as user memo
-			userMemo = combinedMemo
-			chasmMemoPayload = nil
+			customSAs.IndexedFields[name] = payloadValue
 		}
+	}
+	return
+}
 
-		executions = append(executions, &chasm.ExecutionInfo[*commonpb.Payload]{
-			BusinessID:             exec.WorkflowID,
-			RunID:                  exec.RunID,
-			StartTime:              exec.StartTime,
-			CloseTime:              exec.CloseTime,
-			HistoryLength:          exec.HistoryLength,
-			HistorySizeBytes:       exec.HistorySizeBytes,
-			StateTransitionCount:   exec.StateTransitionCount,
-			ChasmSearchAttributes:  chasm.NewSearchAttributesMap(exec.ChasmSearchAttributes),
-			CustomSearchAttributes: exec.SearchAttributes.GetIndexedFields(),
-			Memo:                   userMemo,
-			ChasmMemo:              chasmMemoPayload,
-		})
+// splitUserAndChasmMemo extracts user memo and CHASM memo from the combined memo.
+func (p *visibilityManagerImpl) splitUserAndChasmMemo(exec *store.InternalExecutionInfo) (userMemo *commonpb.Memo, chasmMemoPayload *commonpb.Payload, err error) {
+	combinedMemo, err := deserializeMemo(exec.Memo)
+	if err != nil {
+		return
 	}
 
-	return &chasm.ListExecutionsResponse[*commonpb.Payload]{
-		Executions:    executions,
-		NextPageToken: response.NextPageToken,
-	}, nil
+	if isChasmExecution(exec.SearchAttributes) && combinedMemo != nil {
```

Comment:
nit: do the quicker check first
```suggestion
	if combinedMemo != nil && isChasmExecution(exec.SearchAttributes) {
```